packages/components/src/components/chat/AutoComplete.vue

```diff
@@ -0,0 +1,205 @@
+<template>
+  <v-popper :arrow="false" align="left" placement="top" ref="autocomplete" class="autocomplete">
+    <template #content>
+      <ul data-cy="autocomplete">
+        <li
+          v-for="(command, i) in autoCompletions"
+          :key="i"
+          :class="{ selected: i === selectionIndex }"
+          @click.prevent="onSubmitCompletion(command.name)"
+          @mouseover.prevent="selectionIndex = i"
+        >
+          <div class="command">
+            <b :data-active="selectionIndex === i">{{ command.name }}</b>
+          </div>
+          <span class="description">{{ command.description }}</span>
+        </li>
+      </ul>
+    </template>
+  </v-popper>
+</template>
+
+<script lang="ts">
+import Vue from 'vue';
+import VPopper from '@/components/Popper.vue';
+import { NavieRpc } from '@appland/rpc';
+
+type PopperRef = Vue & { show: () => void; hide: () => void; isVisible: Readonly<boolean> };
+
+export default Vue.extend({
+  name: 'v-auto-complete',
+
+  components: {
+    VPopper,
+  },
+
+  props: {
+    input: {
+      type: String,
+      required: true,
+    },
+    commands: {
+      type: Array as () => NavieRpc.V1.Metadata.Command[],
+      default: () => [],
+    },
+  },
+
+  data() {
+    return {
+      caretOffset: undefined as number | undefined,
+      selectionIndex: 0,
+      onInput: (() => {}) as (e: KeyboardEvent) => void,
+      onCaretChange: () => {},
+    };
+  },
+
+  watch: {
+    autoCompletions(val) {
+      if (val.length) {
+        this.onDisplay();
+      } else {
+        this.onHide();
+      }
+    },
+    input: {
+      immediate: true,
+      handler() {
+        this.onCaretChange();
+      },
+    },
+  },
+
+  computed: {
+    popperRef(): PopperRef | undefined {
+      return this.$refs.autocomplete as PopperRef | undefined;
+    },
+    activeToken(): string | undefined {
+      if (!this.caretOffset || !this.input) return undefined;
+
+      // Don't consider any other token but the first. Commands are always at the beginning of the
+      // input.
+      const firstTokenBoundary = this.input.match(/\s|$/)?.index;
+      if (firstTokenBoundary && this.caretOffset > firstTokenBoundary) return undefined;
+
+      // Find an index to the end of the token the cursor is on
+      const tokenEnd = this.input.slice(this.caretOffset).match(/\s|$/);
+
+      // Calculate the index of the end of the token
+      let endIndex = this.caretOffset;
+      if (tokenEnd) endIndex += tokenEnd?.index ?? 0;
+
+      // activeText is the string from the start of the input to the cursor
+      const activeText = this.input.slice(0, endIndex);
+
+      // We need the word the cursor is on, so we split on whitespace and take the last token
+      const tokens = activeText.split(/\s+/);
+      return tokens[tokens.length - 1];
+    },
+    autoCompletions(): NavieRpc.V1.Metadata.Command[] {
+      if (!this.activeToken) return [];
```

Comment:
Should we disable this feature completely if we don't have an activeToken or the RPC service doesn't support this command?

Reply:
Falling back to an empty array effectively does that here. Nothing is performed if `autoCompletions` is empty. Similarly, if the RPC method doesn't exist, `commands` will fall back to an empty array and effectively be a no-op, aside from computing `activeToken` above.